### PR TITLE
Tracker enhancement

### LINKED ISSUE #46: [BUG] Health, Willpower, Roll Frenzy issues
**Describe the bug**
the latest update seems to have broken the health, and willpower tracks. Not only are the health and willpower tracks filling up backwards (the red "full" boxes should be from right to left, with the blanks/damage starting on the left) but rolling for frenzy also seems broken. It should only roll your remaining willpower (blank undamaged willpower in other words, as per p119 of the VTM rulebook under "tracker and tools") + 1/3 of your humanity... but right now it seems to roll the filled out "red" boxes in your willpower instead (although I think it's rolling the 1/3 humanity ok).  

**Expected behavior**
Health and Willpower tracks should have white boxes equal to your health and willpower values. The remainder should be blacked/redded out. These white boxes should be filled with slashes and Xs starting from the left.

Rolling for Resist Frenzy should only roll dice equal to any blank boxes in your willpower (meaning, no slashes or Xs or black/red) and 1/3 of your humanity.

- FoundryVTT Version 0.7.9
- Foundry-V5 Version 1.3

### PR DESCRIPTION
**Pull request by bmgraves**
Fixes #46 
This change covers: Modifications to the way health/willpower are managed, and Implementation of a migration system for when template.json is updated. 

Health/Willpower trackers.
tracker_example

Health/Willpower trackers now store their Current/Max values, rather then "aggravated" and "superficial" wounds. These are stored as an array of strings which contain either "-" for undamaged, "/" for superficial, or "x" for aggravated. The states themselves are ALSO stored in the template.json. So for future trackers that may want to use this code, they can create their own separate list of states and the code **SHOULD** handle it properly without additional fiddling. (the code looks up what states belong to the tracker it is working on and then grabs the next/previous state from that, rather than having it hard coded in the actor code). Current is simply max - the number of wounds. The boxes are "dumb" IE, you can click them in any order. left clicking moves them forward a step, right clicking cycles them backwards a step. The red/Green Buttons increase/decrease maximum health starting at the RIGHT side of the tracker moving backwards. So any wounds on the last box are deleted as health is reduced.
tracker_example2

Additionally worth noting that the Humanity bar continues to use the old code. Humanity should be redone as well, and the code presented here can serve as a solid baseline to implement that change, but some changes will need to be made to make it function correctly.

Migration System:
System will now store a variable for the systemversion that will be compared when a GM logs into foundry. If the current version of the vtm5e system is newer than the previous version, it will go through the migrations. Currently only one migration exists(for the other change in this pull request) I have started it with a check to look for old values that will be removed, so the change will only occur on actors that need to be updated.

It is worth noting/crediting that the core of the migration system code was copied from the [dnd5e](https://gitlab.com/foundrynet/dnd5e) project written by atropos, with modifications to make it work in this system.

Minor changes:

Some css changes were made to work with the new system. Including using variables for color values in the tracker boxes. Which means theoretically we could implement some future code that would allow tracker boxes to be customized. Probably wont come up, but it is worth noting anyway.

EDIT: There is a minor CSS bug when max health/willpower is put OVER 10. I haven't yet found a way to prevent this. It does not stop any functionality. but it looks bad. I'll keep looking for a solution, but it may be worth looking at before the code is implemented in the next version.

### CODE REVIEW

**Comment by bmgraves** [comment]
Just an adendum to this. I just realized this change breaks the frenzy roll. I somehow forgot to take that into account. I'm working on a fix for this that I'll push into this pull request soon.

**Comment by bmgraves** [comment]
Frenzy roll fixed in last push. Everything should now be working.

**Comment by Rayji96** [comment]
conflicts with the dev branch :(

**Comment by bmgraves** [comment]
Ah dang. I should have probably used the dev branch as my starting branch. I’ll take a look at it tonight or tomorrow and get things sorted out.

File: module/actor/actor-sheet.js [hunk 1]
```diff
@@ -663,6 +668,25 @@ export class VampireActorSheet extends ActorSheet {
     this._vampireRoll(dicePool, this.actor, `${item.data.name}`)
   }
 
+<<<<<<< HEAD
```
Comment (Rayji96) [inline module/actor/actor-sheet.js:671]: remove this line

File: module/actor/actor-sheet.js [hunk 2]
```diff
@@ -663,6 +668,25 @@ export class VampireActorSheet extends ActorSheet {
...
+
+  // There were issues in health tracking from multiple separate fields needing to be updated at once.
+  // The old method "_assignToActorField()" was de-syncing the data. This field does the same thing,
+  // but assigns multiple fields/values at once, before making a single update to the actor data.
+  _updateActorFields (containers) {
+    const actorData = duplicate(this.actor)
+    let i;
+    for (i=0;i < (containers.length);i++) {
```
Comment (Rayji96) [inline module/actor/actor-sheet.js:679]: do `for (let i = 0; i < (containers.length); i++) {`
and remove line 678

File: module/actor/actor-sheet.js [hunk 3]
```diff
@@ -802,6 +828,112 @@ export class VampireActorSheet extends ActorSheet {
     })
   }
 
+  // Increases the number of counters for health/willpower
+  _onPls(event) {
+    event.preventDefault()
+    const element = event.currentTarget
+    const TrackerData = element.parentNode.parentNode.dataset // TrackerData contains the dataset of the "resourceTracker" div that contains all data for tracker management
```
Comment (Rayji96) [inline module/actor/actor-sheet.js:835]: nit but do trackerData
Caps on first letter is reserved for defining classes

File: module/actor/actor-sheet.js [hunk 4]
```diff
@@ -802,6 +828,112 @@ export class VampireActorSheet extends ActorSheet {
...
+    const element = event.currentTarget
+    const dataset = element.dataset
+    const index = Number(element.dataset.index)
+    const fields = dataset.name.split('.')
+    const states = element.parentNode.dataset.states.split(',')
+    const mouse = event.which
+
+    let currentState = dataset.state
+    let stateIndex = states.indexOf(String(currentState))
+    const OldIndex = states.indexOf(String(currentState))
```
Comment (Rayji96) [inline module/actor/actor-sheet.js:896]: nit but do oldIndex 
Caps on first letter is reserved for defining classes

**Comment by Rayji96** [comment]
What is rock.jpg for?

**Comment by Rayji96** [comment]
I'm sorry for the slow review process :(
It's a big change

**Comment by bmgraves** [comment]
> 
> 
> What is rock.jpg for?

I've removed it. There was a point where i was trying to use pieces from another code, and added that to test a bug i was seeing. Apparently I removed the code, and missed that i left that in there. My apologies.

Requested changes have been made. :)

**Comment by tom959** [comment]
Is the review still pending?

**Comment by Rayji96** [comment]
There's a merge conflict. :(
I wrote my own logic for adding and subtracting resource boxes so there'll likely have to be a marriage between the different implementations.

**Comment by bmgraves** [comment]
I'm confused, wasn't this merged, or at least the functionality implemented a while back? Is this merge even still necessary? Whats missing?

**Comment by Rayji96** [comment]
I implemented the + - buttons but didn't implement the way to toggle any box rather than the fixes left to right box counter

**Comment by bmgraves** [comment]
Ah, then this should just be closed. I don't think that my method of tracking was necessarily the best way, I just wanted to resolve the max health/redbox issue.

**Pull request closed without merging**